File: Days 61 - 70/Day 69/LargestProductOfThreeNumbers.cpp
#include <algorithm>
#include <array>
#include <iostream>
#include <limits>
#include <type_traits>
#include <utility>

struct TripleMax
{
	int a = std::numeric_limits<int>::min();
	int b = std::numeric_limits<int>::min();
	int c = std::numeric_limits<int>::min();

	unsigned int totalAdded = 0;

	void Add(int number) noexcept
	{
		if (number > a)
		{
			c = b;
			b = a;
			a = number;
		}
		else if (number > b)
		{
			c = b;
			b = number;
		}
		else if (number > c)
		{
			c = number;
		}

		totalAdded++;
	}
};

void AddToNegativePair(std::pair<int, int>& negativePair, int number) noexcept
{
	if (number < negativePair.first)
	{
		negativePair.second = negativePair.first;
		negativePair.first = number;
	}
	else if (number < negativePair.second)
	{
		negativePair.second = number;
	}
}

template <unsigned int Size, typename = std::enable_if_t<(Size >= 3)>>
int GetMaxProductFromThreeNumbers(const std::array<int, Size>& numbers) noexcept
{
	if (Size == 3)
	{
		return numbers[0] * numbers[1] * numbers[2];
	}

	TripleMax maxPositives;
	std::pair<int, int> minNegatives { std::numeric_limits<int>::max(), std::numeric_limits<int>::max() };
	unsigned int negativeCount = 0;

	for (int number : numbers)
	{
		if (number >= 0)
		{
			maxPositives.Add(number);
		}
		else
		{
			AddToNegativePair(minNegatives, number);
			negativeCount++;
		}
	}

	int productWithNegatives = std::numeric_limits<int>::min();

	if (negativeCount >= 2 && maxPositives.totalAdded >= 1)
	{
		productWithNegatives = minNegatives.first * minNegatives.second * maxPositives.a;
	}

	int productWithoutNegatives = maxPositives.a * maxPositives.b * maxPositives.c;

	return std::max(productWithoutNegatives, productWithNegatives);
}

int main(int argc, char* argv[])
{
	std::cout << GetMaxProductFromThreeNumbers<4>({ -10, -10, 5, 2 }) << "\n";
	std::cout << GetMaxProductFromThreeNumbers<4>({ -10, 10, 5, 2 }) << "\n";
	std::cout << GetMaxProductFromThreeNumbers<6>({ -10, -10, -5, 2, 7, 9 }) << "\n";
	std::cout << GetMaxProductFromThreeNumbers<6>({ -10, 10, -5, 2, 7, 9 }) << "\n";
	std::cout << GetMaxProductFromThreeNumbers<4>({ -10, 5, 0, -20 }) << "\n";
	std::cout << GetMaxProductFromThreeNumbers<4>({ -10, 5, 0, 20 }) << "\n";

	std::cin.get();

	return 0;
}

File: Days 61 - 70/Day 69/LargestProductOfThreeNumbers.cs
using System;

namespace DailyCodingProblem
{
	internal class TripleMax
	{
		public int A { get; private set; } = int.MinValue;
		public int B { get; private set; } = int.MinValue;
		public int C { get; private set; } = int.MinValue;

		public int TotalAdded { get; private set; } = 0;

		public void Add(int number)
		{
			if (number > A)
			{
				C = B;
				B = A;
				A = number;
			}
			else if (number > B)
			{
				C = B;
				B = number;
			}
			else if (number > C)
			{
				C = number;
			}

			TotalAdded++;
		}
	}

	internal class Day69
	{
		private static int Main(string[] args)
		{
			Console.WriteLine(GetMaxProductFromThreeNumbers(new int[] { -10, -10, 5, 2 }));
			Console.WriteLine(GetMaxProductFromThreeNumbers(new int[] { -10, 10, 5, 2 }));
			Console.WriteLine(GetMaxProductFromThreeNumbers(new int[] { -10, -10, -5, 2, 7, 9 }));
			Console.WriteLine(GetMaxProductFromThreeNumbers(new int[] { -10, 10, -5, 2, 7, 9 }));
			Console.WriteLine(GetMaxProductFromThreeNumbers(new int[] { -10, 5, 0, -20 }));
			Console.WriteLine(GetMaxProductFromThreeNumbers(new int[] { -10, 5, 0, 20 }));

			Console.ReadLine();

			return 0;
		}

		private static int GetMaxProductFromThreeNumbers(int[] numbers)
		{
			if (numbers.Length == 3)
			{
				return numbers[0] * numbers[1] * numbers[2];
			}

			TripleMax maxPositives = new TripleMax();
			(int, int) minNegatives = (int.MaxValue, int.MaxValue);
			int negativeCount = 0;

			foreach (int number in numbers)
			{
				if (number >= 0)
				{
					maxPositives.Add(number);
				}
				else
				{
					AddToNegativePair(ref minNegatives, number);
					negativeCount++;
				}
			}

			int productWithNegatives = int.MinValue;

			if (negativeCount >= 2 && maxPositives.TotalAdded >= 1)
			{
				productWithNegatives = minNegatives.Item1 * minNegatives.Item2 * maxPositives.A;
			}

			int productWithoutNegatives = maxPositives.A * maxPositives.B * maxPositives.C;

			return Math.Max(productWithoutNegatives, productWithNegatives);
		}

		private static void AddToNegativePair(ref (int, int) negativePair, int number)
		{
			if (number < negativePair.Item1)
			{
				negativePair.Item2 = negativePair.Item1;
				negativePair.Item1 = number;
			}
			else if (number < negativePair.Item2)
			{
				negativePair.Item2 = number;
			}
		}
	}
}